tests/sm_int/sm_int.c

```diff
@@ -0,0 +1,264 @@
+// Copyright(C) Microsoft Corporation.All rights reserved.
+
+#ifdef __cplusplus
+#include <cinttypes>
+#else
+#include <inttypes.h>
+#include <stdbool.h>
+#endif
+
+#include "windows.h"
+
+#include "testrunnerswitcher.h"
+
+#include "azure_c_util/timer.h"
+#include "azure_c_util/sm.h"
+
+#define ARRAY_SIZE 1000000
+
+/*how many threads maximum. This needs to be slightly higher than the number of CPU threads because we want to see interrupted threads*/
+/*the tests will start from 1*/
+/*for every number of threads (starting from 1... N_MAX_THREADS)*/
+/*the test will spawn 0... N_MAX_THREADS of threads that do "non-barrier" operations*/
+/*the rest of the threads will do "barrier" operations*/
+/*all the threads write in an array of 1.000.000 elements. Each element contains 2 items: 1) a number 2) a type (true = barrier, false = non-barrier)*/
+
+/*there's a shared index the array (starts at zero)*/
+/*there's a shared "source of numbers" (starts at zero)*/
+
+/*a non barrier thread granted execution will interlocked increment the index, interlocked increment the source of numbers and write it*/
+/*a thread granted execution will interlocked increment the index, interlocked increment the source of numbers and write it*/
+
+#define N_MAX_THREADS 32
+
+static volatile LONG barrier_grants = 0;
+static volatile LONG64 barrier_refusals = 0; /*this is just for giggles*/
+
+static volatile LONG non_barrier_grants = 0;
+static volatile LONG64 non_barrier_refusals = 0;
+
+typedef struct ONE_WRITE_TAG
+{
+    LONG what_was_source;
+    bool is_barrier;
+}ONE_WRITE;
+
+typedef struct THREADS_COMMON_TAG
+{
+    volatile LONG nFinishedThreads;
+    SM_HANDLE sm;
+
+    volatile LONG source_of_numbers;
+    volatile LONG current_index;
+    ONE_WRITE writes[ARRAY_SIZE];
+
+    double startTimems;
+
+}THREADS_COMMON;
+
+#define THREAD_TYPE_VALUES \
+    THREAD_BARRIER, \
+    THREAD_NON_BARRIER
+
+MU_DEFINE_ENUM(THREAD_TYPE, THREAD_TYPE_VALUE)
+
+static DWORD WINAPI barrier_thread(
```

Comment:
Can we please use ThreadAPI_ APIs?
They provide the same functionally and we are starting the Linux build work so it is easier if we use the abstractions rather than OS specific APIs.